Update iOS app tutorial (#1639)
bazelbuild/rules_apple: doc/tutorials/ios-app.md

To build applications for Apple devices, Bazel needs to pull the latest
[Apple build rules](https://github.com/bazelbuild/rules_apple)
from its GitHub repository. To enable this, add the following

> [`http_archive`](https://bazel.build/rules/lib/repo/http#http_archive)
> rules to your `WORKSPACE` file:

Reviewer: ```suggestion
statements to your `WORKSPACE` file:
```

There is only one `http_archive` rule, then there is a lot of loads. Do we need to link to `http_archive`?

```starlark
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "build_bazel_rules_apple",
    sha256 = "f003875c248544009c8e8ae03906bbdacb970bc3e5931b40cd76cadeded99632",
    url = "https://github.com/bazelbuild/rules_apple/releases/download/1.1.0/rules_apple.1.1.0.tar.gz",
)

load(
    "@build_bazel_rules_apple//apple:repositories.bzl",
    "apple_rules_dependencies",
)

apple_rules_dependencies()

load(
    "@build_bazel_rules_swift//swift:repositories.bzl",
    "swift_rules_dependencies",
)

swift_rules_dependencies()

load(
    "@build_bazel_rules_swift//swift:extras.bzl",
    "swift_rules_extra_dependencies",
)

swift_rules_extra_dependencies()

load(
    "@build_bazel_apple_support//lib:repositories.bzl",
    "apple_support_dependencies",
)

apple_support_dependencies()
```

> Note: "Always use the

Reviewer: We should remove the quotes around this.

> [latest version of the Apple rules](https://github.com/bazelbuild/rules_apple/releases)
> in the `tag` attribute. Make sure to check the latest dependencies required in
> `rules_apple`'s [project](https://github.com/bazelbuild/rules_apple)."

Reviewer: ```suggestion
[latest version of the Apple rules](https://github.com/bazelbuild/rules_apple/releases)
in the `url` attribute. Make sure to check the latest dependencies required in
`rules_apple`'s [project](https://github.com/bazelbuild/rules_apple)."
```

```bash
INFO: Found 1 target...
Target //ios-app:ios-app up-to-date:
  bazel-out/applebin_ios-ios_sim_arm64-fastbuild-ST-4e6c2a19403f/bin/ios-app/ios-app.ipa
INFO: Elapsed time: 0.141s, Critical Path: 0.00s
```

### Find the build outputs

The `.ipa` file and other outputs are located in the

> `$WORKSPACE/bazel-out/applebin_ios-ios_sim_arm64-fastbuild-ST-4e6c2a19403f/bin/ios-app/ios-app.ipa` directory.

Reviewer: ```suggestion
`$BUILD_WORKSPACE_DIRECTORY/bazel-out/applebin_ios-ios_sim_arm64-fastbuild-ST-4e6c2a19403f/bin/ios-app/ios-app.ipa` directory.
```

Author: Here `$WORKSPACE` is referencing the custom ENV variable set at the beginning of the tutorial, not the Bazel's specific one. I guess they're equivalent, but shouldn't we use the same one for consistency?

Reviewer: Ahh, true.

### Build the app in the simulator

`rules_apple` supports running an app directly in the iOS Simulator.
Replace `build` with `run` in the previous command to both build and
run the application:

```bash
bazel run //ios-app:ios-app
```

> Note: `--ios_simulator_device` and `--ios_simulator_version` control which

Reviewer: ```suggestion
Note: [`--ios_simulator_device`](https://bazel.build/reference/command-line-reference#flag--ios_simulator_device) and [`--ios_simulator_version`](https://bazel.build/reference/command-line-reference#flag--ios_simulator_version) control which
```

version and device will be used when launching the app.

### Generate an Xcode project

There are a few community-provided solutions (such as [rules_xcodeproj](https://github.com/buildbuddy-io/rules_xcodeproj)

> and [Tulsi](https://github.com/bazelbuild/tulsi)) to help generating Xcode 

Reviewer 2: Since Tulsi has a website, should this link there instead of the GitHub repo?

```suggestion
and [Tulsi](https://tulsi.bazel.build/)) to help generating Xcode 
```

> projects. By doing so, you will be able to write, debug and test 

Reviewer: ```suggestion
projects. By doing so, you will be able to write, debug, and test 
```

iOS/macOS/watchOS/tvOS applications as if you were using the Xcode build system.

Let's see how to do so with `rules_xcodeproj`.

Open the `WORKSPACE` file again and add the following:

```starlark
http_archive(
    name = "com_github_buildbuddy_io_rules_xcodeproj",
    sha256 = "c76958e21b7ea48a40fe4130e57911670281f23be93965573b4e90be47d779b4",
    strip_prefix = "rules_xcodeproj-47d9510b64878ae1893d355cbf134d6a4b6eb715",

>     url = "https://github.com/buildbuddy-io/rules_xcodeproj/archive/47d9510b64878ae1893d355cbf134d6a4b6eb715.tar.gz",

Reviewer 2: I assume there's a reason you're pinning to this commit instead of the official 0.7.1 release from a few hours before?

Author: There was an important change to how you can define `top_level_targets` (https://github.com/buildbuddy-io/rules_xcodeproj/pull/965), so I just decided to get ahead and include it here. We can definitely try to keep this up to date and use release versions instead of commits from now on.

Reviewer: Yeah, let's use this commit for now, since it's a big change in how the rule is used. After this we should just use released versions.

)

load(
    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:repositories.bzl",
    "xcodeproj_rules_dependencies",
)

xcodeproj_rules_dependencies()
```

Add the following import at the top of the `BUILD` file:

```starlark
load(
    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
    "top_level_target",
    "xcodeproj",
)
```

We can now define the rule that will generate the Xcode project:

```starlark
xcodeproj(
    name = "xcodeproj",
    project_name = "ios-app",
    tags = ["manual"],
    build_mode = "bazel",
    archived_bundles_allowed = True,
    top_level_targets = [
        top_level_target(":ios-app"),

>     ],

Reviewer: Can you sort these as `buildifier` would?

>     ],

Reviewer: When you aren't specifying `target_environments`, you don't need to use the macro. I would not use the macro here, but then show how to set `"device", "simulator"` down in the `Build the app for a device` section.

Author: Good idea, added.

)
```

To generate the Xcode project, invoke this rule with the following command:

```bash
bazel run //ios-app:xcodeproj
```

> You should be able to open the generated `"ios-app.xcodeproj` and do all the usual

Reviewer: ```suggestion
You should be able to open the generated `ios-app.xcodeproj` (e.g. `xed ios-app.xcodeproj`) and do all the usual
```

To specify a minimum required iOS version, add the `minimum_os_version`
parameter to the `ios_application` build rule in your `BUILD` file.

You should also update the previously defined `xcodeproj` rule to specify
support for building for a device:

```starlark
xcodeproj(
    name = "xcodeproj",
    project_name = "ios-app",
    tags = ["manual"],
    build_mode = "bazel",
    archived_bundles_allowed = True,
    top_level_targets = [
        top_level_target(":ios-app", target_environments = ["device", "simulator"]),

>     ],

Reviewer: Can you fix the ordering of these tags as well?
